### PR TITLE
CloudFront Origin Request Policy

### PR DESCRIPTION
**Pull request by bill-rich**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #14373
Output from acceptance testing:

```
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSCloudFrontOriginRequestPolicy -timeout 120m
=== RUN   TestAccAWSCloudFrontOriginRequestPolicy_basic
=== PAUSE TestAccAWSCloudFrontOriginRequestPolicy_basic
=== RUN   TestAccAWSCloudFrontOriginRequestPolicy_update
=== PAUSE TestAccAWSCloudFrontOriginRequestPolicy_update
=== RUN   TestAccAWSCloudFrontOriginRequestPolicy_noneBehavior
=== PAUSE TestAccAWSCloudFrontOriginRequestPolicy_noneBehavior
=== CONT  TestAccAWSCloudFrontOriginRequestPolicy_basic
=== CONT  TestAccAWSCloudFrontOriginRequestPolicy_noneBehavior
=== CONT  TestAccAWSCloudFrontOriginRequestPolicy_update
--- PASS: TestAccAWSCloudFrontOriginRequestPolicy_noneBehavior (14.52s)
--- PASS: TestAccAWSCloudFrontOriginRequestPolicy_basic (14.60s)
--- PASS: TestAccAWSCloudFrontOriginRequestPolicy_update (24.22s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	24.268s
```

### CODE REVIEW

File: website/docs/r/cloudfront_origin_request_policy.html.markdown [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+* `headers_config` - (Required) An object that determines whether any HTTP headers (and if so, which headers) are included in the origin request key and automatically included in requests that CloudFront sends to the origin. See [Headers Config](#headers-config) for more information.
+* `query_strings_config` - (Required) An object that determines whether any URL query strings in viewer requests (and if so, which query strings) are included in the origin request key and automatically included in requests that CloudFront sends to the origin. See [Query Strings Config](#query-strings-config) for more information.
+
+### Cookies Config
+
+`cookie_behavior` - (Required) Determines whether any cookies in viewer requests are included in the origin request key and automatically included in requests that CloudFront sends to the origin. Valid values are `none`, `whitelist` `all`.
+`cookies` - (Optional) An object that contains a list of cookie names. See [Items](#items) for more information.
```
Comment (YakDriver) [inline website/docs/r/cloudfront_origin_request_policy.html.markdown:54]:
```suggestion
`cookies` - (Optional) Object that contains a list of cookie names. See [Items](#items) for more information.
```

File: website/docs/r/cloudfront_origin_request_policy.html.markdown [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+### Headers Config
+
+`header_behavior` - (Required) Determines whether any HTTP headers are included in the origin request key and automatically included in requests that CloudFront sends to the origin. Valid values are `none`, `whitelist`, `allViewer`, `allViewerAndWhitelistCloudFront`.
+`headers` - (Optional) An object that contains a list of header names. See [Items](#items) for more information.
```
Comment (YakDriver) [inline website/docs/r/cloudfront_origin_request_policy.html.markdown:59]:
```suggestion
`headers` - (Optional) Object that contains a list of header names. See [Items](#items) for more information.
```

File: website/docs/r/cloudfront_origin_request_policy.html.markdown [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+### Query String Config
+
+`query_string_behavior` - (Required) Determines whether any URL query strings in viewer requests are included in the origin request key and automatically included in requests that CloudFront sends to the origin. Valid values are `none`, `whitelist`, `all`.
+`query_strings` - (Optional) An object that contains a list of query string names. See [Items](#items) for more information.
```
Comment (YakDriver) [inline website/docs/r/cloudfront_origin_request_policy.html.markdown:64]:
```suggestion
`query_strings` - (Optional) Object that contains a list of query string names. See [Items](#items) for more information.
```

File: website/docs/r/cloudfront_origin_request_policy.html.markdown [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+### Items
+
+`items` - (Required) A list of item names (cookies, headers, or query strings).
```
Comment (YakDriver) [inline website/docs/r/cloudfront_origin_request_policy.html.markdown:68]:
```suggestion
`items` - (Required) List of item names (cookies, headers, or query strings).
```

File: website/docs/r/cloudfront_origin_request_policy.html.markdown [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `etag` - The current version of the origin request policy.
```
Comment (YakDriver) [inline website/docs/r/cloudfront_origin_request_policy.html.markdown:74]:
```suggestion
* `etag` - Current version of the origin request policy.
```

File: website/docs/r/cloudfront_origin_request_policy.html.markdown [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+* `id` - The identifier for the origin request policy.
```
Comment (YakDriver) [inline website/docs/r/cloudfront_origin_request_policy.html.markdown:75]:
```suggestion
* `id` - Identifier for the origin request policy.
```

File: website/docs/d/cloudfront_origin_request_policy.html.markdown [hunk 7]
```diff
@@ -0,0 +1,54 @@
+---
+subcategory: "CloudFront"
+layout: "aws"
+page_title: "AWS: aws_cloudfront_origin_request_policy"
+description: |-
+  Determines the values that CloudFront includes in requests that it sends to the origin.
```
Comment (YakDriver) [inline website/docs/d/cloudfront_origin_request_policy.html.markdown:6]:
```suggestion
  Provides information that CloudFront includes in requests that it sends to the origin.
```

File: website/docs/d/cloudfront_origin_request_policy.html.markdown [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+---
+
+# Data Source: aws_cloudfront_origin_request_policy
+
```
Comment (YakDriver) [inline website/docs/d/cloudfront_origin_request_policy.html.markdown:10]:
```suggestion

Provides an origin request policy, and other metadata, for Amazon CloudFront.

```

File: website/docs/d/cloudfront_origin_request_policy.html.markdown [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+## Example Usage
+
+The following example below creates a CloudFront origin request policy.
```
Comment (YakDriver) [inline website/docs/d/cloudfront_origin_request_policy.html.markdown:13]:
```suggestion
The following example gets an CloudFront origin request policy.
```

File: website/docs/r/cloudfront_origin_request_policy.html.markdown [hunk 10]
```diff
@@ -0,0 +1,75 @@
+---
+subcategory: "CloudFront"
+layout: "aws"
+page_title: "AWS: aws_cloudfront_origin_request_policy"
+description: |-
+  Determines the values that CloudFront includes in requests that it sends to the origin.
```
Comment (YakDriver) [inline website/docs/r/cloudfront_origin_request_policy.html.markdown:6]:
```suggestion
  Manages an origin request policy for Amazon CloudFront.
```

File: website/docs/r/cloudfront_origin_request_policy.html.markdown [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+---
+
+# Resource: aws_cloudfront_origin_request_policy
+
```
Comment (YakDriver) [inline website/docs/r/cloudfront_origin_request_policy.html.markdown:10]:
```suggestion

Manages an origin request policy for Amazon CloudFront. After you create an origin request policy, you can attach it to one or more cache behaviors. When it’s attached to a cache behavior, the origin request policy determines the values that CloudFront includes in requests that it sends to the origin.

```

File: website/docs/d/cloudfront_origin_request_policy.html.markdown [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+### Headers Config
+
+`header_behavior` - Determines whether any HTTP headers are included in the origin request key and automatically included in requests that CloudFront sends to the origin. Valid values are `none`, `whitelist`, `allViewer`, `allViewerAndWhitelistCloudFront`.
+`headers` - An object that contains a list of header names. See [Items](#items) for more information.
+
+### Query String Config
+
+`query_string_behavior` - Determines whether any URL query strings in viewer requests are included in the origin request key and automatically included in requests that CloudFront sends to the origin. Valid values are `none`, `whitelist`, `all`.
+`query_strings` - An object that contains a list of query string names. See [Items](#items) for more information.
```
Comment (YakDriver) [inline website/docs/d/cloudfront_origin_request_policy.html.markdown:50]:
```suggestion
`query_strings` - Object that contains a list of query string names. See [Items](#items) for more information.
```

File: website/docs/d/cloudfront_origin_request_policy.html.markdown [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+### Cookies Config
+
+`cookie_behavior` - Determines whether any cookies in viewer requests are included in the origin request key and automatically included in requests that CloudFront sends to the origin. Valid values are `none`, `whitelist` `all`.
+`cookies` - An object that contains a list of cookie names. See [Items](#items) for more information.
+
+### Headers Config
+
+`header_behavior` - Determines whether any HTTP headers are included in the origin request key and automatically included in requests that CloudFront sends to the origin. Valid values are `none`, `whitelist`, `allViewer`, `allViewerAndWhitelistCloudFront`.
+`headers` - An object that contains a list of header names. See [Items](#items) for more information.
```
Comment (YakDriver) [inline website/docs/d/cloudfront_origin_request_policy.html.markdown:45]:
```suggestion
`headers` - Object that contains a list of header names. See [Items](#items) for more information.
```

File: website/docs/d/cloudfront_origin_request_policy.html.markdown [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+* `etag` - The current version of the origin request policy.
+* `headers_config` - An object that determines whether any HTTP headers (and if so, which headers) are included in the origin request key and automatically included in requests that CloudFront sends to the origin. See [Headers Config](#headers-config) for more information.
+* `query_strings_config` - An object that determines whether any URL query strings in viewer requests (and if so, which query strings) are included in the origin request key and automatically included in requests that CloudFront sends to the origin. See [Query Strings Config](#query-strings-config) for more information.
+
+### Cookies Config
+
+`cookie_behavior` - Determines whether any cookies in viewer requests are included in the origin request key and automatically included in requests that CloudFront sends to the origin. Valid values are `none`, `whitelist` `all`.
+`cookies` - An object that contains a list of cookie names. See [Items](#items) for more information.
```
Comment (YakDriver) [inline website/docs/d/cloudfront_origin_request_policy.html.markdown:40]:
```suggestion
`cookies` - Object that contains a list of cookie names. See [Items](#items) for more information.
```

File: website/docs/d/cloudfront_origin_request_policy.html.markdown [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+* `name` - A unique name to identify the origin request policy.
+* `id` - The identifier for the origin request policy.
+
+## Attributes Reference
+
+* `comment` - A comment to describe the origin request policy.
+* `cookies_config` - An object that determines whether any cookies in viewer requests (and if so, which cookies) are included in the origin request key and automatically included in requests that CloudFront sends to the origin. See [Cookies Config](#cookies-config) for more information.
+* `etag` - The current version of the origin request policy.
+* `headers_config` - An object that determines whether any HTTP headers (and if so, which headers) are included in the origin request key and automatically included in requests that CloudFront sends to the origin. See [Headers Config](#headers-config) for more information.
+* `query_strings_config` - An object that determines whether any URL query strings in viewer requests (and if so, which query strings) are included in the origin request key and automatically included in requests that CloudFront sends to the origin. See [Query Strings Config](#query-strings-config) for more information.
```
Comment (YakDriver) [inline website/docs/d/cloudfront_origin_request_policy.html.markdown:35]:
```suggestion
* `query_strings_config` - Object that determines whether any URL query strings in viewer requests (and if so, which query strings) are included in the origin request key and automatically included in requests that CloudFront sends to the origin. See [Query Strings Config](#query-strings-config) for more information.
```

File: website/docs/d/cloudfront_origin_request_policy.html.markdown [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+
+* `name` - A unique name to identify the origin request policy.
+* `id` - The identifier for the origin request policy.
+
+## Attributes Reference
+
+* `comment` - A comment to describe the origin request policy.
+* `cookies_config` - An object that determines whether any cookies in viewer requests (and if so, which cookies) are included in the origin request key and automatically included in requests that CloudFront sends to the origin. See [Cookies Config](#cookies-config) for more information.
+* `etag` - The current version of the origin request policy.
+* `headers_config` - An object that determines whether any HTTP headers (and if so, which headers) are included in the origin request key and automatically included in requests that CloudFront sends to the origin. See [Headers Config](#headers-config) for more information.
```
Comment (YakDriver) [inline website/docs/d/cloudfront_origin_request_policy.html.markdown:34]:
```suggestion
* `headers_config` - Object that determines whether any HTTP headers (and if so, which headers) are included in the origin request key and automatically included in requests that CloudFront sends to the origin. See [Headers Config](#headers-config) for more information.
```

File: website/docs/d/cloudfront_origin_request_policy.html.markdown [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+
+The following arguments are supported:
+
+* `name` - A unique name to identify the origin request policy.
+* `id` - The identifier for the origin request policy.
+
+## Attributes Reference
+
+* `comment` - A comment to describe the origin request policy.
+* `cookies_config` - An object that determines whether any cookies in viewer requests (and if so, which cookies) are included in the origin request key and automatically included in requests that CloudFront sends to the origin. See [Cookies Config](#cookies-config) for more information.
```
Comment (YakDriver) [inline website/docs/d/cloudfront_origin_request_policy.html.markdown:32]:
```suggestion
* `cookies_config` - Object that determines whether any cookies in viewer requests (and if so, which cookies) are included in the origin request key and automatically included in requests that CloudFront sends to the origin. See [Cookies Config](#cookies-config) for more information.
```

File: website/docs/d/cloudfront_origin_request_policy.html.markdown [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - A unique name to identify the origin request policy.
+* `id` - The identifier for the origin request policy.
+
+## Attributes Reference
+
+* `comment` - A comment to describe the origin request policy.
```
Comment (YakDriver) [inline website/docs/d/cloudfront_origin_request_policy.html.markdown:31]:
```suggestion
* `comment` - Comment to describe the origin request policy.
```

**Review by YakDriver — CHANGES_REQUESTED** [review]
Looks good. Just a few minor changes needed.

File: aws/data_source_aws_cloudfront_origin_request_policy.go [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+		},
+	}
+}
+
+func dataSourceAwsCloudFrontOriginRequestPolicyRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).cloudfrontconn
+
+	if d.Get("id").(string) == "" {
+		if err := dataSourceAwsCloudFrontOriginRequestPolicyFindByName(d, conn); err != nil {
+			return fmt.Errorf("Unable to find origin request policy by name: %s", err.Error())
```
Comment (YakDriver) [inline aws/data_source_aws_cloudfront_origin_request_policy.go:116]:
```suggestion
			return fmt.Errorf("unable to find origin request policy by name: %s", err.Error())
```

File: aws/data_source_aws_cloudfront_origin_request_policy.go [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+	}
+
+	if d.Id() != "" {
+		request := &cloudfront.GetOriginRequestPolicyInput{
+			Id: aws.String(d.Id()),
+		}
+
+		resp, err := conn.GetOriginRequestPolicy(request)
+		if err != nil {
+			return fmt.Errorf("Unable to retrieve origin request policy with ID %s: %s", d.Id(), err.Error())
```
Comment (YakDriver) [inline aws/data_source_aws_cloudfront_origin_request_policy.go:127]:
```suggestion
			return fmt.Errorf("unable to retrieve origin request policy with ID %s: %s", d.Id(), err.Error())
```

**Review by YakDriver — APPROVED** [review]
This is looking good. Few minors things left (especially `d.IsNewResource()`) and it's good to go!

GovCloud:
```
--- SKIP: TestAccAWSCloudFrontOriginRequestPolicy_basic (1.16s)
--- SKIP: TestAccAWSCloudFrontOriginRequestPolicy_noneBehavior (1.16s)
--- SKIP: TestAccAWSCloudFrontOriginRequestPolicy_update (1.16s)
```

`us-west-2`:
```
--- PASS: TestAccAWSCloudFrontOriginRequestPolicy_basic (10.76s)
--- PASS: TestAccAWSCloudFrontOriginRequestPolicy_noneBehavior (10.77s)
--- PASS: TestAccAWSCloudFrontOriginRequestPolicy_update (17.53s)
```

File: aws/resource_aws_cloudfront_distribution.go [hunk 21]
```diff
@@ -151,6 +151,10 @@ func resourceAwsCloudFrontDistribution() *schema.Resource {
 							Optional: true,
 							Default:  0,
 						},
+						"origin_request_policy_id": {
```
Comment (vaskevich) [inline aws/resource_aws_cloudfront_distribution.go:154]: It seems that this new parameter was never added to the documentation.

**Pull request merged**